File: surface/widgets/nativescript/components/widgets/concepts/formal-verification/StatusGrid.ts
import {
  StackLayout,
  GridLayout,
  Label,
  Button,
  ScrollView,
  WrapLayout,
  Color,
  View,
} from '@nativescript/core';

export type StatusGridState = 'idle' | 'cellHovered' | 'cellSelected';
export type StatusGridEvent =
  | { type: 'HOVER_CELL' }
  | { type: 'CLICK_CELL' }
  | { type: 'SORT' }
  | { type: 'FILTER' }
  | { type: 'LEAVE_CELL' }
  | { type: 'DESELECT' };

export function statusGridReducer(state: StatusGridState, event: StatusGridEvent): StatusGridState {
  switch (state) {
    case 'idle':
      if (event.type === 'HOVER_CELL') return 'cellHovered';
      if (event.type === 'CLICK_CELL') return 'cellSelected';
      if (event.type === 'SORT') return 'idle';
      if (event.type === 'FILTER') return 'idle';
      return state;
    case 'cellHovered':
      if (event.type === 'LEAVE_CELL') return 'idle';
      if (event.type === 'CLICK_CELL') return 'cellSelected';
      return state;
    case 'cellSelected':
      if (event.type === 'DESELECT') return 'idle';
      if (event.type === 'CLICK_CELL') return 'cellSelected';
      return state;
    default:
      return state;
  }
}

export type CellStatus = 'passed' | 'failed' | 'running' | 'pending' | 'timeout';
export type StatusFilterValue = 'all' | 'passed' | 'failed';

export interface StatusGridItem {
  id: string;
  name: string;
  status: CellStatus;
  duration?: number;
}

export interface StatusGridProps {
  items: StatusGridItem[];
  columns?: number;
  showAggregates?: boolean;
  variant?: 'compact' | 'expanded';
  onCellSelect?: (item: StatusGridItem) => void;
  filterStatus?: StatusFilterValue;
}

const STATUS_COLORS: Record<CellStatus, string> = {
  passed: '#22c55e',
  failed: '#ef4444',
  running: '#3b82f6',
  pending: '#9ca3af',
  timeout: '#f97316',
};

const STATUS_LABELS: Record<CellStatus, string> = {
  passed: 'Passed',
  failed: 'Failed',
  running: 'Running',
  pending: 'Pending',
  timeout: 'Timeout',
};

function formatDuration(ms: number): string {
  if (ms < 1000) return `${ms}ms`;
  return `${(ms / 1000).toFixed(1)}s`;
}

export function createStatusGrid(props: StatusGridProps): { view: View; dispose: () => void } {
  let state: StatusGridState = 'idle';
  let filter: StatusFilterValue = props.filterStatus ?? 'all';
  let selectedIndex: number | null = null;
  const disposers: (() => void)[] = [];
  const isCompact = (props.variant ?? 'expanded') === 'compact';

  function send(event: StatusGridEvent) {
    state = statusGridReducer(state, event);
  }

  const root = new StackLayout();
  root.className = 'clef-status-grid';
  root.automationText = 'Verification status matrix';

  function getFiltered(): StatusGridItem[] {
    if (filter === 'all') return props.items;
    return props.items.filter((i) => i.status === filter);
  }

  function render() {
    root.removeChildren();
    const filtered = getFiltered();

    // Summary counts
    if (props.showAggregates !== false) {
      const counts: Record<CellStatus, number> = { passed: 0, failed: 0, running: 0, pending: 0, timeout: 0 };
      for (const item of props.items) counts[item.status]++;
      const parts: string[] = [];
      if (counts.passed > 0) parts.push(`${counts.passed} passed`);
      if (counts.failed > 0) parts.push(`${counts.failed} failed`);
      if (counts.running > 0) parts.push(`${counts.running} running`);
      if (counts.pending > 0) parts.push(`${counts.pending} pending`);
      if (counts.timeout > 0) parts.push(`${counts.timeout} timeout`);

      const summaryLabel = new Label();
      summaryLabel.text = parts.join(', ');
      summaryLabel.fontSize = isCompact ? 12 : 14;
      summaryLabel.padding = '8 12';
      root.addChild(summaryLabel);
    }

    // Filter bar
    const filterBar = new StackLayout();
    filterBar.orientation = 'horizontal';
    filterBar.padding = '4 12';

    for (const value of ['all', 'passed', 'failed'] as StatusFilterValue[]) {
      const btn = new Button();
      btn.text = value.charAt(0).toUpperCase() + value.slice(1);
      btn.fontSize = isCompact ? 11 : 13;
      btn.padding = isCompact ? '2 8' : '4 12';
      btn.borderWidth = 1;
      btn.borderColor = filter === value ? new Color('#6366f1') : new Color('#d1d5db');
      btn.borderRadius = 4;
      btn.backgroundColor = filter === value ? new Color('#eef2ff') : new Color('transparent');
      btn.fontWeight = filter === value ? 'bold' : 'normal';
      btn.marginRight = 4;

      const handler = () => {
        filter = value;
        selectedIndex = null;
        send({ type: 'FILTER' });
        render();
      };
      btn.on('tap', handler);
      disposers.push(() => btn.off('tap', handler));
      filterBar.addChild(btn);
    }
    root.addChild(filterBar);

    // Grid cells
    const grid = new WrapLayout();
    grid.orientation = 'horizontal';
    grid.padding = '4 12';

    filtered.forEach((item, index) => {
      const cell = new StackLayout();
      cell.width = isCompact ? 80 : 120;
      cell.padding = isCompact ? '4' : '8 12';
      cell.borderRadius = 4;
      cell.margin = '2';
      cell.borderWidth = 2;
      cell.borderColor = selectedIndex === index ? new Color('#6366f1') : new Color('transparent');

      // Status dot
      const dot = new Label();
      dot.text = '\u25CF';
      dot.color = new Color(STATUS_COLORS[item.status]);
      dot.fontSize = isCompact ? 10 : 14;
      cell.addChild(dot);

      // Name
      const nameLabel = new Label();
      nameLabel.text = item.name;
      nameLabel.fontSize = isCompact ? 10 : 12;
      cell.addChild(nameLabel);

      // Duration
      if (!isCompact && item.duration != null) {
        const durLabel = new Label();
        durLabel.text = formatDuration(item.duration);
        durLabel.fontSize = 11;
        durLabel.color = new Color('#6b7280');
        cell.addChild(durLabel);
      }

      cell.automationText = `${item.name}: ${STATUS_LABELS[item.status]}${item.duration != null ? `, ${formatDuration(item.duration)}` : ''}`;

      const cellHandler = () => {
        selectedIndex = index;
        send({ type: 'CLICK_CELL' });
        props.onCellSelect?.(item);
        render();
      };
      cell.on('tap', cellHandler);
      disposers.push(() => cell.off('tap', cellHandler));

      grid.addChild(cell);
    });
    root.addChild(grid);

    // Detail panel for selected cell
    if (selectedIndex != null && filtered[selectedIndex]) {
      const item = filtered[selectedIndex];
      const detail = new StackLayout();
      detail.padding = '12';
      detail.borderTopWidth = 1;
      detail.borderTopColor = new Color('#e5e7eb');

      const nameLabel = new Label();
      nameLabel.text = item.name;
      nameLabel.fontWeight = 'bold';
      nameLabel.fontSize = 14;
      detail.addChild(nameLabel);

      const statusLabel = new Label();
      statusLabel.text = `Status: ${STATUS_LABELS[item.status]}`;
      statusLabel.fontSize = 13;
      detail.addChild(statusLabel);

      if (item.duration != null) {
        const durLabel = new Label();
        durLabel.text = `Duration: ${formatDuration(item.duration)}`;
        durLabel.fontSize = 13;
        durLabel.color = new Color('#6b7280');
        detail.addChild(durLabel);
      }

      root.addChild(detail);
    }
  }

  render();

  return {
    view: root,
    dispose() { disposers.forEach((d) => d()); },
  };
}

export default createStatusGrid;

File: surface/widgets/appkit/components/widgets/concepts/formal-verification/StatusGrid.swift
import AppKit

// MARK: - Status grid types

enum CellStatus: String {
    case passed, failed, running, pending, timeout

    var color: NSColor {
        switch self {
        case .passed: return NSColor(red: 0.133, green: 0.773, blue: 0.369, alpha: 1)
        case .failed: return NSColor(red: 0.937, green: 0.267, blue: 0.267, alpha: 1)
        case .running: return NSColor(red: 0.231, green: 0.510, blue: 0.965, alpha: 1)
        case .pending: return NSColor(red: 0.612, green: 0.639, blue: 0.686, alpha: 1)
        case .timeout: return NSColor(red: 0.976, green: 0.451, blue: 0.086, alpha: 1)
        }
    }

    var displayLabel: String { rawValue.capitalized }
}

struct StatusGridItem {
    let id: String
    let name: String
    let status: CellStatus
    var duration: Int?
}

enum StatusFilterValue: String, CaseIterable {
    case all, passed, failed
}

// MARK: - StatusGridView

class StatusGridView: NSView {

    enum State: String { case idle; case cellHovered; case cellSelected }

    private var widgetState: State = .idle
    private var items: [StatusGridItem] = []
    private var columns: Int = 4
    private var showAggregates: Bool = true
    private var compact: Bool = false
    private var filter: StatusFilterValue = .all
    private var hoveredIndex: Int?
    private var selectedIndex: Int?
    private var focusIndex: Int = 0
    private var trackingArea: NSTrackingArea?

    var onCellSelect: ((StatusGridItem) -> Void)?

    private let rootStack = NSStackView()
    private let summaryLabel = NSTextField(labelWithString: "")
    private let filterBar = NSStackView()
    private let gridContainer = NSStackView()
    private let tooltipField = NSTextField(labelWithString: "")
    private let detailPanel = NSStackView()
    private let detailName = NSTextField(labelWithString: "")
    private let detailStatus = NSTextField(labelWithString: "")
    private let detailDuration = NSTextField(labelWithString: "")

    // MARK: - State machine

    func send(_ event: String, index: Int? = nil) {
        switch widgetState {
        case .idle:
            if event == "HOVER_CELL" { widgetState = .cellHovered; hoveredIndex = index }
            if event == "CLICK_CELL" { widgetState = .cellSelected; selectedIndex = index; focusIndex = index ?? 0 }
            if event == "SORT" || event == "FILTER" { /* stays idle */ }
        case .cellHovered:
            if event == "LEAVE_CELL" { widgetState = .idle; hoveredIndex = nil }
            if event == "CLICK_CELL" { widgetState = .cellSelected; selectedIndex = index; focusIndex = index ?? 0 }
        case .cellSelected:
            if event == "DESELECT" { widgetState = .idle; selectedIndex = nil }
            if event == "CLICK_CELL" { selectedIndex = index; focusIndex = index ?? 0 }
        }
        rebuildGrid()
    }

    // MARK: - Init

    override init(frame: NSRect) {
        super.init(frame: frame)
        setupSubviews()
        setAccessibilityRole(.group)
        setAccessibilityLabel("Matrix grid displaying verification statuses")
    }

    required init?(coder: NSCoder) { super.init(coder: coder) }

    // MARK: - Setup

    private func setupSubviews() {
        wantsLayer = true

        rootStack.orientation = .vertical
        rootStack.spacing = 6
        rootStack.translatesAutoresizingMaskIntoConstraints = false
        addSubview(rootStack)
        NSLayoutConstraint.activate([
            rootStack.topAnchor.constraint(equalTo: topAnchor, constant: 8),
            rootStack.leadingAnchor.constraint(equalTo: leadingAnchor, constant: 8),
            rootStack.trailingAnchor.constraint(equalTo: trailingAnchor, constant: -8),
            rootStack.bottomAnchor.constraint(equalTo: bottomAnchor, constant: -8),
        ])

        summaryLabel.font = .systemFont(ofSize: 13)
        rootStack.addArrangedSubview(summaryLabel)

        filterBar.orientation = .horizontal
        filterBar.spacing = 4
        for fv in StatusFilterValue.allCases {
            let btn = NSButton(title: fv.rawValue.capitalized, target: self, action: #selector(filterClicked(_:)))
            btn.bezelStyle = .roundRect
            btn.tag = StatusFilterValue.allCases.firstIndex(of: fv) ?? 0
            btn.setAccessibilityLabel("Filter \(fv.rawValue)")
            filterBar.addArrangedSubview(btn)
        }
        rootStack.addArrangedSubview(filterBar)

        gridContainer.orientation = .vertical
        gridContainer.spacing = 4
        rootStack.addArrangedSubview(gridContainer)

        tooltipField.font = .systemFont(ofSize: 12)
        tooltipField.wantsLayer = true
        tooltipField.layer?.backgroundColor = NSColor(white: 0.12, alpha: 0.95).cgColor
        tooltipField.textColor = .white
        tooltipField.isBezeled = false
        tooltipField.isEditable = false
        tooltipField.drawsBackground = true
        tooltipField.isHidden = true
        addSubview(tooltipField)

        detailPanel.orientation = .vertical
        detailPanel.spacing = 4
        detailPanel.isHidden = true
        detailName.font = .systemFont(ofSize: 14, weight: .semibold)
        detailPanel.addArrangedSubview(detailName)
        detailStatus.font = .systemFont(ofSize: 13)
        detailPanel.addArrangedSubview(detailStatus)
        detailDuration.font = .systemFont(ofSize: 13)
        detailDuration.textColor = .secondaryLabelColor
        detailPanel.addArrangedSubview(detailDuration)
        rootStack.addArrangedSubview(detailPanel)

        setupTrackingArea()
    }

    // MARK: - Configure

    func configure(items: [StatusGridItem], columns: Int = 4, showAggregates: Bool = true, compact: Bool = false) {
        self.items = items
        self.columns = columns
        self.showAggregates = showAggregates
        self.compact = compact
        rebuildGrid()
    }

    private var filteredItems: [StatusGridItem] {
        if filter == .all { return items }
        return items.filter { $0.status.rawValue == filter.rawValue }
    }

    private func formatDuration(_ ms: Int) -> String {
        ms < 1000 ? "\(ms)ms" : String(format: "%.1fs", Double(ms) / 1000.0)
    }

    private func rebuildGrid() {
        gridContainer.arrangedSubviews.forEach { $0.removeFromSuperview() }

        let filtered = filteredItems
        let actualCols = min(columns, filtered.count)
        guard actualCols > 0 else { return }

        // Summary
        var counts = [CellStatus: Int]()
        for item in items { counts[item.status, default: 0] += 1 }
        var parts = [String]()
        if let c = counts[.passed], c > 0 { parts.append("\(c) passed") }
        if let c = counts[.failed], c > 0 { parts.append("\(c) failed") }
        if let c = counts[.running], c > 0 { parts.append("\(c) running") }
        if let c = counts[.pending], c > 0 { parts.append("\(c) pending") }
        if let c = counts[.timeout], c > 0 { parts.append("\(c) timeout") }
        summaryLabel.stringValue = parts.joined(separator: ", ")
        summaryLabel.isHidden = !showAggregates

        // Grid rows
        var row = NSStackView()
        row.orientation = .horizontal
        row.spacing = compact ? 2 : 4
        row.distribution = .fillEqually

        for (index, item) in filtered.enumerated() {
            if index > 0 && index % actualCols == 0 {
                gridContainer.addArrangedSubview(row)
                row = NSStackView()
                row.orientation = .horizontal
                row.spacing = compact ? 2 : 4
                row.distribution = .fillEqually
            }

            let cell = NSStackView()
            cell.orientation = .vertical
            cell.spacing = compact ? 2 : 4
            cell.alignment = compact ? .centerX : .leading
            cell.wantsLayer = true
            cell.layer?.cornerRadius = 4

            if index == selectedIndex {
                cell.layer?.borderWidth = 2
                cell.layer?.borderColor = NSColor.controlAccentColor.cgColor
            } else if index == hoveredIndex {
                cell.layer?.borderWidth = 1
                cell.layer?.borderColor = NSColor.separatorColor.cgColor
            }

            // Status dot
            let dot = NSView()
            dot.wantsLayer = true
            dot.layer?.backgroundColor = item.status.color.cgColor
            dot.layer?.cornerRadius = compact ? 5 : 7
            dot.translatesAutoresizingMaskIntoConstraints = false
            cell.addArrangedSubview(dot)
            let dotSize: CGFloat = compact ? 10 : 14
            NSLayoutConstraint.activate([
                dot.widthAnchor.constraint(equalToConstant: dotSize),
                dot.heightAnchor.constraint(equalToConstant: dotSize),
            ])

            let nameField = NSTextField(labelWithString: item.name)
            nameField.font = .systemFont(ofSize: compact ? 10 : 12)
            nameField.lineBreakMode = .byTruncatingTail
            nameField.maximumNumberOfLines = 1
            cell.addArrangedSubview(nameField)

            if !compact, let dur = item.duration {
                let durField = NSTextField(labelWithString: formatDuration(dur))
                durField.font = .systemFont(ofSize: 11)
                durField.textColor = .secondaryLabelColor
                cell.addArrangedSubview(durField)
            }

            cell.setAccessibilityLabel("\(item.name): \(item.status.displayLabel)\(item.duration != nil ? ", \(formatDuration(item.duration!))" : "")")
            cell.tag = index

            let click = NSClickGestureRecognizer(target: self, action: #selector(cellClicked(_:)))
            cell.addGestureRecognizer(click)

            row.addArrangedSubview(cell)
        }
        if !row.arrangedSubviews.isEmpty { gridContainer.addArrangedSubview(row) }

        // Detail panel
        if let si = selectedIndex, si < filtered.count {
            let item = filtered[si]
            detailPanel.isHidden = false
            detailName.stringValue = item.name
            detailStatus.stringValue = "Status: \(item.status.displayLabel)"
            if let dur = item.duration {
                detailDuration.stringValue = "Duration: \(formatDuration(dur))"
                detailDuration.isHidden = false
            } else { detailDuration.isHidden = true }
        } else { detailPanel.isHidden = true }

        // Tooltip
        if widgetState == .cellHovered, let hi = hoveredIndex, hi < filtered.count {
            let item = filtered[hi]
            var text = "\(item.name) \u{2014} \(item.status.displayLabel)"
            if let dur = item.duration { text += " (\(formatDuration(dur)))" }
            tooltipField.stringValue = text
            tooltipField.isHidden = false
            tooltipField.sizeToFit()
        } else { tooltipField.isHidden = true }
    }

    // MARK: - Actions

    @objc private func filterClicked(_ sender: NSButton) {
        filter = StatusFilterValue.allCases[sender.tag]
        selectedIndex = nil
        hoveredIndex = nil
        focusIndex = 0
        send("FILTER")
    }

    @objc private func cellClicked(_ recognizer: NSClickGestureRecognizer) {
        guard let cell = recognizer.view as? NSStackView else { return }
        let index = cell.tag
        send("CLICK_CELL", index: index)
        let filtered = filteredItems
        if index < filtered.count { onCellSelect?(filtered[index]) }
    }

    // MARK: - Keyboard

    override var acceptsFirstResponder: Bool { true }

    override func keyDown(with event: NSEvent) {
        let filtered = filteredItems
        guard !filtered.isEmpty else { super.keyDown(with: event); return }
        let actualCols = min(columns, filtered.count)

        switch event.keyCode {
        case 124: focusIndex = min(focusIndex + 1, filtered.count - 1) // Right
        case 123: focusIndex = max(focusIndex - 1, 0) // Left
        case 125: focusIndex = min(focusIndex + actualCols, filtered.count - 1) // Down
        case 126: focusIndex = max(focusIndex - actualCols, 0) // Up
        case 36: send("CLICK_CELL", index: focusIndex); if focusIndex < filtered.count { onCellSelect?(filtered[focusIndex]) }
        case 53: send("DESELECT")
        default: super.keyDown(with: event)
        }
    }

    // MARK: - Mouse tracking

    private func setupTrackingArea() {
        let area = NSTrackingArea(rect: bounds, options: [.mouseEnteredAndExited, .mouseMoved, .activeInKeyWindow, .inVisibleRect], owner: self, userInfo: nil)
        addTrackingArea(area)
        trackingArea = area
    }

    override func updateTrackingAreas() {
        super.updateTrackingAreas()
        if let ta = trackingArea { removeTrackingArea(ta) }
        setupTrackingArea()
    }

    override func mouseMoved(with event: NSEvent) {
        let point = gridContainer.convert(event.locationInWindow, from: nil)
        for row in gridContainer.arrangedSubviews {
            guard let rowStack = row as? NSStackView else { continue }
            for cell in rowStack.arrangedSubviews {
                if cell.frame.contains(gridContainer.convert(point, to: rowStack)) {
                    send("HOVER_CELL", index: cell.tag)
                    return
                }
            }
        }
        send("LEAVE_CELL")
    }

    override func mouseExited(with event: NSEvent) { send("LEAVE_CELL") }

    deinit { if let ta = trackingArea { removeTrackingArea(ta) } }
}
